### PR TITLE
Add DroppedSpansCallback to batch span processor to detect dropped spans.

### PR DESCRIPTION
**Pull request by collinvandyck**
We're using the batch span processor, but we wanted to have visibility into when the span channel was full, resulting in dropped spans. This PR adds the ability to configure the processor with a callback that will get invoked when the batch span processor cannot send on the spans channel.

### CODE REVIEW

File: sdk/trace/batch_span_processor.go [hunk 1]
```diff
@@ -378,6 +389,7 @@ func (bsp *batchSpanProcessor) enqueueBlockOnQueueFull(ctx context.Context, sd R
 		return true
 	default:
 		atomic.AddUint32(&bsp.dropped, 1)
+		bsp.o.DroppedSpansCallback()
```
Comment (Aneurysm9) [inline sdk/trace/batch_span_processor.go:392]:
```suggestion
		go bsp.o.DroppedSpansCallback()
```

This should not allow the callback to block the processor.

**Review by Aneurysm9 — COMMENTED** [review]
This seems reasonable, though I might prefer that we expose the existing dropped span count as a metric when the metric API is stable and usable here.  If we do go this route then the callback cannot be allowed to block the processor and this will need some tests.

File: sdk/trace/batch_span_processor.go (hunk 1 above)
Comment (collinvandyck) [reply sdk/trace/batch_span_processor.go]: Thanks for the suggestion. Addressed in https://github.com/open-telemetry/opentelemetry-go/pull/2766/commits/7357a7f6d43e81067d30db228853d851c6aae263.

File: sdk/trace/batch_span_processor.go [hunk 2]
```diff
@@ -378,6 +388,9 @@ func (bsp *batchSpanProcessor) enqueueBlockOnQueueFull(ctx context.Context, sd R
 		return true
 	default:
 		atomic.AddUint32(&bsp.dropped, 1)
+		if bsp.o.DroppedSpansCallback != nil {
+			go bsp.o.DroppedSpansCallback()
```
Comment (jmacd) [inline sdk/trace/batch_span_processor.go:392]: I would prefer this to be a blocking call, one the user is responsible for making very quickly. I also want to see OTel specify a conventional metric to count these drops, but don't want to block you here.

**Review by MrAlias — CHANGES_REQUESTED** [review]
This will change the `BatchSpanProcessorOptions` type to no longer be comparable. This will be a backwards incompatible change for a exported type marked stable. I do not think we can accept this change without violating our stability guarantees.

**Comment by collinvandyck** [comment]
Hmm if `NewBatchSpanProcessor` were to return an exported concrete type instead of the `SpanProcessor` interface, I could add a `MetricsSnapshot` method to it that would return the number of dropped spans, but it seems like this kind of work would likely collide with the work being done on the metrics API.

I'm a little stuck on how to get at this `dropped` field otherwise. Kind of seems like we may not have a way forward with this PR unless anyone else has ideas?

**Comment by jmacd** [comment]
> anyone else has ideas?

I suggest we create a new type for dropped span errors:

```
type DroppedSpanError struct {
    DroppedSpans int
}
```

Then use `otel.Handle(DroppedSpanError{DroppedSpans: N})` to record N dropped spans. This will allow a future integration between the metrics SDK and the trace SDK to export the count of dropped spans w/o any new interfaces.

**Comment by jmacd** [comment]
> Is there a reason we wouldn't have the BSP contain it's own metric instruments to track and export these values?

That's a specification waiting to be written. I believe no one has worked on this because the metrics SDK wasn't stabilized, but now's the time.

**Comment by MrAlias** [comment]
> > Is there a reason we wouldn't have the BSP contain it's own metric instruments to track and export these values?
> 
> That's a specification waiting to be written. I believe no one has worked on this because the metrics SDK wasn't stabilized, but now's the time.

I have a strong preference to got with this solution to the problem! I think we could get a PoC PR or branch created to showcase this and make a proposal to the specification.

**Comment by collinvandyck** [comment]
I think this one probably can't go any further so I'm closing it out. Looking forward to the future metrics API!

**Pull request closed without merging**